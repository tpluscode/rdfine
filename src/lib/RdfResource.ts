import { Term, NamedNode, DatasetCore, BlankNode } from 'rdf-js'
import cf, { SingleContextClownface } from 'clownface'
import ns from '@rdfjs/namespace'
import once from 'once'

const rdf = ns('http://www.w3.org/1999/02/22-rdf-syntax-ns#')

export interface RdfResource<D extends DatasetCore = DatasetCore> {
  readonly id: Term | null
  readonly types: Term[]
  readonly _node: SingleContextClownface<D>
  hasType (type: string | NamedNode): boolean
}

export default class RdfResourceImpl<D extends DatasetCore = DatasetCore> implements RdfResource<D> {
  public readonly _node: SingleContextClownface<D>
  private readonly __initializeProperties: (() => void)
  public static __ns?: any
  private static __defaults = new Map<string, unknown>()

  public constructor(node: SingleContextClownface<D> | { dataset: D; term: NamedNode | BlankNode; graph?: NamedNode }) {
    const contexts = cf(node).toArray()

    if (contexts.length !== 1) {
      throw new Error('RdfResource can only be initialized from a single node. Got ' + contexts.length)
    }

    this._node = cf(node).toArray()[0]

    this.__initializeProperties = once(() => {
      const self = this as any
      RdfResourceImpl.__defaults.forEach((value, key) => {
        if (typeof self[key] === 'undefined') {
          self[key] = value
        }
      })
    })

    this.__initializeProperties()
  }

  public get id() {
    if (!this._node.term) throw new Error('Graph context does not represent a single node')

    return this._node.term
  }

  public get types() {
    return this._node.out(rdf.type).terms
  }

  public hasType(type: string | NamedNode) {
    const typeNode = typeof type === 'string' ? this._node.namedNode(type) : type

    return this._node.has(rdf.type, typeNode).terms.length > 0
  }
}
