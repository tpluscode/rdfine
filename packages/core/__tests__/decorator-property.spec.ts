import { prefixes } from '@zazuko/rdf-vocabularies'
import cf from 'clownface'
import { namespace, property, crossBoundaries } from '../index'
import RdfResource from '../RdfResource'
import { parse, vocabs } from './_helpers'
import { DatasetCore, DefaultGraph, Literal, NamedNode, Term } from 'rdf-js'
import { literal, namedNode } from '@rdfjs/data-model'
import rdfExt from 'rdf-ext'
import DatasetExt from 'rdf-ext/lib/Dataset'

const { ex, foaf, schema, rdf } = vocabs

describe('decorator', () => {
  describe('term', () => {
    describe('getter', () => {
      it('returns raw named node from property', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix foaf: <${prefixes.foaf}> .

          ex:res foaf:friend ex:friend .`)
        class Resource extends RdfResource {
          @property({ path: foaf.friend })
          friend?: NamedNode
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })
        const friend = instance.friend

        // then
        expect(friend!.value).toEqual(ex.friend.value)
      })

      it('returns raw blank node from property', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix foaf: <${prefixes.foaf}> .
          @prefix schema: <${prefixes.schema}> .
          
          ex:res ex:likes [
            a ex:Blank
          ] .
        `)
        class Resource extends RdfResource {
          @property({ path: ex.likes })
          likes?: Term
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // then
        expect(instance.likes!.termType).toEqual('BlankNode')
      })

      it('returns raw literal node from property', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix foaf: <${prefixes.foaf}> .
          @prefix schema: <${prefixes.schema}> .
          
          ex:res foaf:name "John Doe"@en-us .
        `)
        class Resource extends RdfResource {
          @property({ path: foaf.name })
          name?: Literal
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })
        const name = instance.name

        // then
        expect(name!.value).toEqual('John Doe')
        expect(name!.language).toEqual('en-us')
      })

      it('throws for multiple values', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix foaf: <${prefixes.foaf}> .
          @prefix schema: <${prefixes.schema}> .
          
          ex:res schema:children ex:Hansel, ex:Gretel .
        `)
        class Resource extends RdfResource {
          @property({ path: schema.children })
          children?: Literal
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // then
        expect(() => instance.children).toThrow()
      })

      it('return array when annotated', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix schema: <${prefixes.schema}> .
          
          ex:res schema:children ex:Hansel, ex:Gretel .
        `)
        class Resource extends RdfResource {
          @property({ path: schema.children, values: 'array' })
          children?: Term[]
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // then
        expect(instance.children).toHaveLength(2)
        expect(instance.children!.map(c => c.value)).toContain(ex.Hansel.value)
        expect(instance.children!.map(c => c.value)).toContain(ex.Gretel.value)
      })

      it('return correct node when annotated with namespace, using prop name', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix foaf: <${prefixes.foaf}> .

          ex:res foaf:friend ex:friend .`)

        @namespace(foaf)
        class Resource extends RdfResource {
          @property()
          friend?: NamedNode
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })
        const friend = instance.friend

        // then
        expect(friend!.value).toEqual(ex.friend.value)
      })

      it('get resource by paths of arbitrary length', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix foaf: <${prefixes.foaf}> .

          ex:res foaf:friend ex:friend .
          ex:friend ex:worksAt ex:factory .
          ex:factory foaf:name "RDF Software House" .`)

        @namespace(foaf)
        class Resource extends RdfResource {
          @property({
            path: [
              'friend',
              'http://example.com/worksAt',
              foaf.name,
            ],
          })
          friendsWorkplaceName?: Literal
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // then
        expect(instance.friendsWorkplaceName!.value).toEqual('RDF Software House')
      })

      it('return correct node when annotated with namespace', async () => {
      // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix foaf: <${prefixes.foaf}> .

          ex:res foaf:friend ex:friend .`)

        @namespace(foaf)
        class Resource extends RdfResource {
          @property({ path: 'friend' })
          colleague?: NamedNode
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })
        const friend = instance.colleague

        // then
        expect(friend!.value).toEqual(ex.friend.value)
      })

      it('strict throws when object not found', async () => {
        // given
        const dataset = rdfExt.dataset()
        class Resource extends RdfResource {
          @property({ path: foaf.name, strict: true })
          name!: Literal
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })
        const getName = () => instance.name

        // then
        expect(getName).toThrow()
      })

      it('strict can be initialized', async () => {
        // given
        const dataset = rdfExt.dataset()
        class Resource extends RdfResource {
          @property.literal({ path: foaf.name, strict: true, initial: 'bar' })
          foo!: string
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // then
        expect(instance.foo).toEqual('bar')
        expect(dataset.toCanonical()).toMatchSnapshot()
      })

      it('array can be initialized', async () => {
        // given
        const dataset = rdfExt.dataset()
        class Resource extends RdfResource {
          @property.literal({ path: foaf.name, values: 'array', initial: 'bar' })
          foo!: string
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // then
        expect(instance.foo).toEqual(
          expect.arrayContaining(['bar']),
        )
        expect(dataset.toCanonical()).toMatchSnapshot()
      })

      it('array can be initialized with multiple values', async () => {
        // given
        const dataset = rdfExt.dataset()
        class Resource extends RdfResource {
          @property.literal({ path: foaf.name, values: 'array', initial: ['foo', 'bar'] })
          foo!: string
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // then
        expect(instance.foo).toEqual(
          expect.arrayContaining(['foo', 'bar']),
        )
        expect(dataset.toCanonical()).toMatchSnapshot()
      })

      it('rdf list can be initialized', async () => {
        // given
        const dataset = rdfExt.dataset()
        class Resource extends RdfResource {
          @property.literal({ path: foaf.name, values: 'list', initial: 'bar' })
          foo!: string
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // then
        expect(instance.foo).toEqual(
          expect.arrayContaining(['bar']),
        )
        expect(dataset.toCanonical()).toMatchSnapshot()
      })

      it('rdf list can be initialized with multiple values', async () => {
        // given
        const dataset = rdfExt.dataset()
        class Resource extends RdfResource {
          @property.literal({ path: foaf.name, values: 'list', initial: ['foo', 'bar'] })
          foo!: string
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // then
        expect(instance.foo).toEqual(
          expect.arrayContaining(['foo', 'bar']),
        )
        expect(dataset.toCanonical()).toMatchSnapshot()
      })

      it('returns rdf list array', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix schema: <${prefixes.schema}> .
          
          ex:res ex:letters ( "a" "b" "c" ) .
        `)
        class Resource extends RdfResource {
          @property({ path: ex.letters, values: 'list' })
          letters!: Literal[]
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // then
        expect(instance.letters.map(l => l.value)).toEqual(['a', 'b', 'c'])
      })

      it('returns empty rdf list array', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix schema: <${prefixes.schema}> .
          
          ex:res ex:letters ( ) .
        `)
        class Resource extends RdfResource {
          @property({ path: ex.letters, values: 'list' })
          letters!: Literal[]
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // then
        expect(instance.letters).toEqual([])
      })

      it('throws when rdf list is the object but not annotated', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix foaf: <${prefixes.foaf}> .
          
          ex:res foaf:knows ( ex:Will ex:Joe ex:Sindy ) .
        `)
        class Resource extends RdfResource {
          @property({ path: foaf.knows })
          friend!: Term
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // then
        expect(() => instance.friend).toThrow()
      })
    })

    describe('setter', () => {
      it('can replace literal with blank', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix foaf: <${prefixes.foaf}> .
          @prefix schema: <${prefixes.schema}> .
          
          ex:res foaf:name "John Doe"@en-us .
          [
            a ex:BlankNodeName ;
            ex:first "John" ;
            ex:last "Doe"
          ] .
        `)
        class Resource extends RdfResource {
          @property({ path: foaf.name })
          name?: Term | null
        }

        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // when
        instance.name = cf({ dataset }).has(rdf.type, ex.BlankNodeName).term

        // then
        expect(dataset.toCanonical()).toMatchSnapshot()
      })

      it('can set empty array, removing objects', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix foaf: <${prefixes.foaf}> .
          @prefix schema: <${prefixes.schema}> .
          
          ex:res foaf:name "John"@en-us, "Johann"@de, "Jan"@pl, "Jean"@fr .
        `)
        class Resource extends RdfResource {
          @property({ path: foaf.name, values: 'array' })
          name?: Term[]
        }

        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // when
        instance.name = []

        // then
        expect(dataset.toCanonical()).toMatchSnapshot()
      })

      it('can set array', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix foaf: <${prefixes.foaf}> .
          @prefix schema: <${prefixes.schema}> .
          
          ex:res foaf:name "John"@en-us, "Johann"@de .
        `)
        class Resource extends RdfResource {
          @property({ path: foaf.name, values: 'array' })
          name?: Term[]
        }

        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // when
        instance.name = [
          literal('Jan', 'pl'),
          literal('Jean', 'fr'),
        ]

        // then
        expect(dataset.toCanonical()).toMatchSnapshot()
      })

      it('throws when setting array to non-array getter', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix foaf: <${prefixes.foaf}> .
          @prefix schema: <${prefixes.schema}> .
          
          ex:res foaf:name "John"@en-us, "Johann"@de, "Jan"@pl, "Jean"@fr .
        `)
        class Resource extends RdfResource {
          @property({ path: foaf.name })
          name?: any
        }

        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // when
        const setArray = () => { instance.name = [] }

        // then
        expect(setArray).toThrow()
      })

      it('sets value at paths of arbitrary length', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix foaf: <${prefixes.foaf}> .

          ex:res foaf:friend ex:friend .
          ex:friend ex:worksAt ex:factory .
          ex:factory foaf:name "RDF Software House" .`)

        @namespace(foaf)
        class Resource extends RdfResource {
          @property({
            path: [
              'friend',
              'http://example.com/worksAt',
              foaf.name,
            ],
          })
          friendsWorkplaceName?: Literal
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })
        instance.friendsWorkplaceName = literal('Google')

        // then
        expect(dataset.toCanonical()).toMatchSnapshot()
      })

      it('sets nil for empty rdf list', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix foaf: <${prefixes.foaf}> .
          
          ex:res foaf:knows ( ex:Will ex:Joe ex:Sindy ) .
        `)
        class Resource extends RdfResource {
          @property({ path: foaf.knows, values: 'list' })
          friend!: Term[]
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })
        instance.friend = []

        // then
        expect(dataset.toCanonical()).toMatchSnapshot()
      })

      it('setting null to rdf list removes triple', async () => {
        // given
        const dataset = await parse(`
          @prefix ex: <${prefixes.ex}> .
          @prefix foaf: <${prefixes.foaf}> .
          
          ex:res foaf:knows ( ex:Will ex:Joe ex:Sindy ) .
        `)
        class Resource extends RdfResource {
          @property({ path: foaf.knows, values: 'array' })
          friend!: Term[] | null
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })
        instance.friend = null

        // then
        expect(dataset.toCanonical()).toMatchSnapshot()
      })
    })

    describe('initial', () => {
      it('sets initial value from node', async () => {
        // given
        const dataset = rdfExt.dataset()
        class Resource extends RdfResource {
          @property({
            path: schema.name,
            initial: literal('foo'),
          })
          name!: Literal
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // then
        expect(instance.name.value).toEqual('foo')
        expect(dataset.toCanonical()).toMatchSnapshot()
      })

      it('sets initial value from clownface object', async () => {
        // given
        const dataset = rdfExt.dataset()
        class Resource extends RdfResource {
          @property({
            path: schema.name,
            initial: (self: Resource) => self._selfGraph.blankNode('foo'),
          })
          name!: Literal
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // then
        expect(instance.name.value).toEqual('foo')
        expect(dataset.toCanonical()).toMatchSnapshot()
      })

      it('sets initial value from function', async () => {
        // given
        const dataset = rdfExt.dataset()
        class Resource extends RdfResource {
          @property({
            path: schema.name,
            initial: (self: Resource) => namedNode(`${self.id.value}/child`),
          })
          child!: NamedNode
        }

        // when
        const instance = new Resource({
          dataset,
          term: ex.res,
        })

        // then
        expect(instance.child.value).toEqual('http://example.com/res/child')
        expect(dataset.toCanonical()).toMatchSnapshot()
      })
    })

    describe('over multiple graphs', () => {
      class Resource<D extends DatasetCore> extends RdfResource<D> {
        @property({ path: foaf.knows })
        friend!: Term

        @property({
          path: crossBoundaries(foaf.knows),
        })
        allFriendsAcross!: Term[]

        @property(({
          path: foaf.knows,
          subjectFromAllGraphs: true,
        }))
        allKnownFriends!: Term[]
      }

      function namedGraphTests(newResource: (dataset: DatasetExt, term: NamedNode, graph?: NamedNode | DefaultGraph) => Resource<DatasetExt>) {
        describe('getter', () => {
          it('returns value from same graph', async () => {
            // given
            const dataset = await parse(`
              @prefix ex: <${prefixes.ex}> .
              @prefix foaf: <${prefixes.foaf}> .
              
              ex:John foaf:knows ex:Will ex:John .
              ex:John foaf:knows ex:Sindy ex:Sindy .
              ex:John foaf:knows ex:Brad .
            `)

            // when
            const instance = newResource(dataset, ex.John, ex.John)

            // then
            expect(instance.allKnownFriends).toEqual(
              expect.arrayContaining([ex.Will, ex.Sindy, ex.Brad]),
            )
          })

          it('returns self when dataset is empty', async () => {
            // given
            const dataset = rdfExt.dataset()

            // when
            const instance = newResource(dataset, ex.John)

            // then
            expect(instance.allKnownFriends).toEqual([])
          })

          it('returns value from default graph if unspecified', async () => {
            // given
            const dataset = await parse(`
              @prefix ex: <${prefixes.ex}> .
              @prefix foaf: <${prefixes.foaf}> .
              
              ex:John foaf:knows ex:Will ex:John .
              ex:John foaf:knows ex:Sindy ex:Sindy .
              ex:John foaf:knows ex:Brad .
            `)

            // when
            const instance = newResource(dataset, ex.John)

            // then
            expect(instance.friend.value).toEqual(ex.Brad.value)
          })

          it('ignores crossingBoundaries option', async () => {
            // given
            const dataset = await parse(`
              @prefix ex: <${prefixes.ex}> .
              @prefix foaf: <${prefixes.foaf}> .
              
              ex:John foaf:knows ex:Will ex:John .
              ex:John foaf:knows ex:Will ex:Sindy .
              ex:John foaf:knows ex:Will .
            `)

            // when
            const instance = newResource(dataset, ex.John, ex.John)

            // then
            expect(instance.allFriendsAcross).toHaveLength(1)
          })

          it('de-duplicates terms', async () => {
            // given
            const dataset = await parse(`
              @prefix ex: <${prefixes.ex}> .
              @prefix foaf: <${prefixes.foaf}> .
              
              ex:John foaf:knows ex:Will ex:John .
              ex:John foaf:knows ex:Will ex:Sindy .
              ex:John foaf:knows ex:Will .
            `)

            // when
            const instance = newResource(dataset, ex.John, ex.John)

            // then
            expect(instance.allKnownFriends).toHaveLength(1)
          })
        })

        describe('setter', () => {
          it('sets value to same graph', async () => {
            // given
            const dataset = await parse(`
              @prefix ex: <${prefixes.ex}> .
              @prefix foaf: <${prefixes.foaf}> .
              
              ex:John foaf:knows ex:Will ex:John .
              ex:John foaf:knows ex:Sindy ex:Sindy .
              ex:John foaf:knows ex:Brad .
            `)
            const instance = newResource(dataset, ex.John, ex.John)

            // when
            instance.friend = ex.Holly

            // then
            expect(instance._unionGraph.dataset.toCanonical()).toMatchSnapshot()
          })

          it('sets value to default graph if unspecified', async () => {
            // given
            const dataset = await parse(`
              @prefix ex: <${prefixes.ex}> .
              @prefix foaf: <${prefixes.foaf}> .
              
              ex:John foaf:knows ex:Will ex:John .
              ex:John foaf:knows ex:Sindy ex:Sindy .
              ex:John foaf:knows ex:Brad .
            `)
            const instance = newResource(dataset, ex.John)

            // when
            instance.friend = ex.Holly

            // then
            expect(instance._unionGraph.dataset.toCanonical()).toMatchSnapshot()
          })
        })

        describe('initial', () => {
          it('sets initial values in self graph', () => {
            // given
            const dataset = rdfExt.dataset()
            const instance = newResource(dataset, ex.John, ex.John)

            // when
            instance.friend = ex.Holly

            // then
            expect(instance._unionGraph.dataset.toCanonical()).toMatchSnapshot()
          })
        })
      }

      describe('constructed from plain object', () => {
        namedGraphTests((dataset, term, graph) => {
          return new Resource({ dataset, term, graph })
        })
      })

      describe('constructed from clownface graph', () => {
        namedGraphTests((dataset, term, graph) => {
          return new Resource(cf({ dataset, term, graph }))
        })
      })
    })
  })
})
