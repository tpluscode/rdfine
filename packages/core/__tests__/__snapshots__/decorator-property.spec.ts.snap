// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`decorator term getter array can be initialized with multiple values: toMatchSnapshot 1`] = `
"<http://example.com/res> <http://xmlns.com/foaf/0.1/name> \\"bar\\" .
<http://example.com/res> <http://xmlns.com/foaf/0.1/name> \\"foo\\" .
"
`;

exports[`decorator term getter array can be initialized: toMatchSnapshot 1`] = `
"<http://example.com/res> <http://xmlns.com/foaf/0.1/name> \\"bar\\" .
"
`;

exports[`decorator term getter rdf list can be initialized with multiple values: toMatchSnapshot 1`] = `
"<http://example.com/res> <http://xmlns.com/foaf/0.1/name> _:c14n0 .
_:c14n0 <http://www.w3.org/1999/02/22-rdf-syntax-ns#first> \\"foo\\" .
_:c14n0 <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest> _:c14n1 .
_:c14n1 <http://www.w3.org/1999/02/22-rdf-syntax-ns#first> \\"bar\\" .
_:c14n1 <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest> <http://www.w3.org/1999/02/22-rdf-syntax-ns#nil> .
"
`;

exports[`decorator term getter rdf list can be initialized: toMatchSnapshot 1`] = `
"<http://example.com/res> <http://xmlns.com/foaf/0.1/name> _:c14n0 .
_:c14n0 <http://www.w3.org/1999/02/22-rdf-syntax-ns#first> \\"bar\\" .
_:c14n0 <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest> <http://www.w3.org/1999/02/22-rdf-syntax-ns#nil> .
"
`;

exports[`decorator term getter strict can be initialized: toMatchSnapshot 1`] = `
"<http://example.com/res> <http://xmlns.com/foaf/0.1/name> \\"bar\\" .
"
`;

exports[`decorator term initial sets initial value from clownface object: toMatchSnapshot 1`] = `
"<http://example.com/res> <http://schema.org/name> _:c14n0 .
"
`;

exports[`decorator term initial sets initial value from function: toMatchSnapshot 1`] = `
"<http://example.com/res> <http://schema.org/name> <http://example.com/res/child> .
"
`;

exports[`decorator term initial sets initial value from node: toMatchSnapshot 1`] = `
"<http://example.com/res> <http://schema.org/name> \\"foo\\" .
"
`;

exports[`decorator term over multiple graphs constructed from clownface graph initial sets initial values in self graph: toMatchSnapshot 1`] = `
"<http://example.com/John> <http://xmlns.com/foaf/0.1/knows> <http://example.com/Holly> <http://example.com/John> .
"
`;

exports[`decorator term over multiple graphs constructed from clownface graph setter sets value to default graph if unspecified: toMatchSnapshot 1`] = `
"<http://example.com/John> <http://xmlns.com/foaf/0.1/knows> <http://example.com/Holly> .
<http://example.com/John> <http://xmlns.com/foaf/0.1/knows> <http://example.com/Sindy> <http://example.com/Sindy> .
<http://example.com/John> <http://xmlns.com/foaf/0.1/knows> <http://example.com/Will> <http://example.com/John> .
"
`;

exports[`decorator term over multiple graphs constructed from clownface graph setter sets value to same graph: toMatchSnapshot 1`] = `
"<http://example.com/John> <http://xmlns.com/foaf/0.1/knows> <http://example.com/Brad> .
<http://example.com/John> <http://xmlns.com/foaf/0.1/knows> <http://example.com/Holly> <http://example.com/John> .
<http://example.com/John> <http://xmlns.com/foaf/0.1/knows> <http://example.com/Sindy> <http://example.com/Sindy> .
"
`;

exports[`decorator term setter can replace literal with blank: toMatchSnapshot 1`] = `
"<http://example.com/res> <http://xmlns.com/foaf/0.1/name> _:c14n0 .
_:c14n0 <http://example.com/first> \\"John\\" .
_:c14n0 <http://example.com/last> \\"Doe\\" .
_:c14n0 <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://example.com/BlankNodeName> .
"
`;

exports[`decorator term setter can set array: toMatchSnapshot 1`] = `
"<http://example.com/res> <http://xmlns.com/foaf/0.1/name> \\"Jan\\"@pl .
<http://example.com/res> <http://xmlns.com/foaf/0.1/name> \\"Jean\\"@fr .
"
`;

exports[`decorator term setter can set empty array, removing objects: toMatchSnapshot 1`] = `""`;

exports[`decorator term setter sets list when property allows list or single and setting array with multiple items: toMatchSnapshot 1`] = `
"<http://example.com/res> <http://example.com/foo> _:c14n0 .
<http://example.com/res> <http://xmlns.com/foaf/0.1/knows> <http://www.w3.org/1999/02/22-rdf-syntax-ns#nil> .
_:c14n0 <http://www.w3.org/1999/02/22-rdf-syntax-ns#first> \\"bar\\" .
_:c14n0 <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest> _:c14n1 .
_:c14n1 <http://www.w3.org/1999/02/22-rdf-syntax-ns#first> \\"baz\\" .
_:c14n1 <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest> <http://www.w3.org/1999/02/22-rdf-syntax-ns#nil> .
"
`;

exports[`decorator term setter sets nil for empty rdf list: toMatchSnapshot 1`] = `
"<http://example.com/res> <http://xmlns.com/foaf/0.1/knows> <http://www.w3.org/1999/02/22-rdf-syntax-ns#nil> .
"
`;

exports[`decorator term setter sets value at paths of arbitrary length: toMatchSnapshot 1`] = `
"<http://example.com/factory> <http://xmlns.com/foaf/0.1/name> \\"Google\\" .
<http://example.com/friend> <http://example.com/worksAt> <http://example.com/factory> .
<http://example.com/res> <http://xmlns.com/foaf/0.1/friend> <http://example.com/friend> .
"
`;

exports[`decorator term setter setting null to rdf list removes triple: toMatchSnapshot 1`] = `""`;
